File: 0994-rotting-oranges/0994-rotting-oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int m = (int)grid.size();
        int n = (int)grid[0].size();
        int fresh = 0;
        queue<pair<int,int>> q;
        for (int r=0; r<m; r++){
            for (int c=0; c<n; c++){
                if (grid[r][c] == 1)
                    fresh++;
                else if (grid[r][c] == 2)
                    q.push({r,c});
            }
        }
        int minutes = 0;
        int dr[4] = {0,-1,0,1};
        int dc[4] = {-1,0,1,0};
        while (!q.empty()){
            int sz = (int)q.size();
            bool madeRotten = false;
            for (int i=0; i<sz; i++){
                auto [r,c] = q.front();
                q.pop();
                for (int k=0; k<4; k++){
                    int nr = r+dr[k];
                    int nc = c+dc[k];
                    if (0 <= nr && nr < m &&
                        0 <= nc && nc < n &&
                        grid[nr][nc] == 1){
                            grid[nr][nc] = 2;
                            q.push({nr,nc});
                            fresh--;
                            madeRotten = true;
                        }
                }
            }
            if (madeRotten)
                minutes++;
        }
        return fresh == 0 ? minutes : -1;
    }
};

File: 0994-rotting-oranges/0994-rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        fresh = 0
        q = deque()
        for r in range(m):
            for c in range(n):
                if grid[r][c] == 2:
                    q.append((r,c))
                elif grid[r][c] == 1:
                    fresh += 1
        dr = (0,-1,0,1)
        dc = (-1,0,1,0)
        minutes = 0
        while len(q) != 0:
            madeRotten = False
            sz = len(q)
            for _ in range(sz):
                (r,c) = q.popleft()
                for k in range(4):
                    nr = r+dr[k]
                    nc = c+dc[k]
                    if (0 <= nr and nr < m and
                        0 <= nc and nc < n and 
                        grid[nr][nc] == 1):
                        grid[nr][nc] = 2
                        q.append((nr,nc))
                        fresh -= 1
                        madeRotten = True
            if madeRotten:
                minutes += 1
        return minutes if fresh == 0 else -1